reflex/event.py

```diff
@@ -392,21 +397,62 @@ def call_event_handler(event_handler: EventHandler, arg: Var) -> EventSpec:
 
     Args:
         event_handler: The event handler.
-        arg: The argument to pass to the event handler.
+        arg_spec: The lambda that define the argument(s) to pass to the event handler.
+
+    Raises:
+        ValueError: if number of arguments expected by event_handler doesn't match the spec.
 
     Returns:
         The event spec from calling the event handler.
     """
     args = inspect.getfullargspec(event_handler.fn).args
+
+    # handle new API using lambda to define triggers
+    if isinstance(arg_spec, ArgsSpec):
+        parsed_args = parse_args_spec(arg_spec)
+
+        if len(args) == len(["self", *parsed_args]):
+            return event_handler(*parsed_args)  # type: ignore
+        else:
+            source = inspect.getsource(arg_spec)
+            raise ValueError(
+                f"number of arguments in {event_handler.fn.__name__} "
+                f"doesn't match the definition '{source.strip().strip(',')}'"
+            )
+    else:
+        console.deprecate(
+            feature_name="EVENT_ARG API for triggers",
+            reason="Replaced by new API using lambda allow arbitrary number of args",
+            deprecation_version="0.2.8",
+            removal_version="0.2.9",
+        )
     if len(args) == 1:
         return event_handler()
     assert (
         len(args) == 2
     ), f"Event handler {event_handler.fn} must have 1 or 2 arguments."
-    return event_handler(arg)
+    return event_handler(arg_spec)
 
 
-def call_event_fn(fn: Callable, arg: Var) -> list[EventSpec]:
+def parse_args_spec(arg_spec: ArgsSpec):
+    """Parse the args provided in the ArgsSpec of an event trigger.
+
+    Args:
+        arg_spec: The spec of the args.
+
+    Returns:
+        The parsed args.
+    """
+    spec = inspect.getfullargspec(arg_spec)
+    return arg_spec(
+        *[
+            BaseVar(name=f"_{l_arg}", type_=FrontendEvent, is_local=True)
```

Comment:
```suggestion
            BaseVar(
                name=f"_{l_arg}",
                type_=spec.annotations.get(l_arg, FrontendEvent),
                is_local=True,
            )
```

We should let the user supply the correct types here:

```python
class C1State(State):
    def mock_handler(self, _e, bravo, charlie):
        pass


def test_component_event_trigger_arbitrary_args():
    class Obj(rx.Base):
        custom: int = 0

    def on_foo(
        _e,
        alpha: str,
        bravo: dict[str, Any],
        charlie: Obj,
    ):
        return [_e.target.value, bravo["nested"], charlie.custom + 42]

    class C1(Component):
        library = "/local"
        tag = "C1"

        def get_event_triggers(self) -> Dict[str, Any]:
            return {
                "on_foo": on_foo,
            }

    rendered = C1.create(on_foo=C1State.mock_handler).render()
    assert rendered["props"][0] == (
        "onFoo={(__e,_alpha,_bravo,_charlie) => addEvents("
        '[Event("c1_state.mock_handler", {_e:__e.target.value,bravo:_bravo["nested"],charlie:(_charlie.custom + 42)})], '
        "(__e,_alpha,_bravo,_charlie))}"
    )
```

Reply:
I've added your suggestion, it seems to work when running a normal app, but running `pytest` fail with type errors.
(for both the Dict and custom Obj arguments)